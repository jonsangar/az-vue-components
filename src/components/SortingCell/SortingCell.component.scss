.az-sorting-cell{
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 65px;
  margin-bottom: 0;
  overflow: hidden;
  background-image: linear-gradient(270deg, $color-gray-start 0%, $color-gray-end 100%) !important;
  border-radius: 0;
  box-shadow: inset 1px 2px 2px rgba(255,255,255,.4);

  &--header{
    background-image: linear-gradient(270deg, $color-gray7 0%, white 100%) !important;
    -webkit-box-shadow: 0 11px 8px -5px rgba(0,0,0,0.3);
    box-shadow: 0 11px 8px -5px rgba(0,0,0,0.3);
  }

  &::before{
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 5px;
    content: '';
    background-image: linear-gradient(180deg, rgba(31, 41, 51, 0) 50%, rgba(31, 41, 51, 0.5) 100%);
    opacity: 0.6;
  }

  // .handle{
  //   position: absolute;
  //   width: 70%;
  //   height: 100%;
  // }

  .handleIcon{
    position: absolute;
    top: 0;
    bottom: 0;
    left: calc(#{half()} + #{quarter()});
    margin: auto;
    opacity: .5;
  }

  .handle, .handle--header {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    z-index: 5;
    width: 50%;
    height: double();
    padding-left: calc(#{half()} + #{quarter()});
    margin: auto;
    line-height: double();
    color: color(gray6);
    text-align: left;
    cursor: pointer;
  }



  .editButton {
    position: absolute;
    top: 0;
    right: half();
    bottom: 0;
    z-index: 10;
    width: double();
    height: double();
    margin: auto;
    line-height: double();
    color: color(gray4);
    text-align: center;
    cursor: pointer;
    background-color: color(gray2);
    border-radius: 5px;

  }

  &--header {
    .editButton{
      color: color(gray4);
      background-color: color(gray8);
    }
  }

  &__content{
    position: absolute;
    width: 100%;
    height: 100%;

    &__title {
      @extend %h5-bold-white;

      position: absolute;
      top: 0;
      bottom: 0;
      z-index: 15;
      width: 100%;
      height: fit-content;
      margin: auto;
      margin-left: half();
      font-weight: bold !important;
      // line-height: 1rem;
      -webkit-touch-callout: none; /* iOS Safari */
      -webkit-user-select: none; /* Safari */
      -khtml-user-select: none; /* Konqueror HTML */
      -moz-user-select: none; /* Old versions of Firefox */
      -ms-user-select: none; /* Internet Explorer/Edge */
      user-select: none; /* Non-prefixed version, currently
                          supported by Chrome, Edge, Opera and Firefox */


      .az-sorting-cell--header &{
        color: color(gray2) !important;
      }


      & &--input{
        @extend %h5-bold-white;

        position: relative;
        z-index: 10;
        padding-left: 0;
        font-weight: bold !important;
        line-height: 1rem;
        background-color: transparent;
        border: none;

        .az-sorting-cell--header &{
          color: color(gray2) !important;
        }

      }

      & &--input[disabled="disabled"]{
          z-index: 0;
      }
    }

    &__subtitle{
      position: absolute;
      bottom: 0;
      left: .3rem;
      z-index: 100;
      width: 105%;
      overflow: hidden;
      font-size: font-size(small1);
      color: color(danger);
      text-overflow: ellipsis;
      white-space: nowrap;

      &::before{
        content: '*';
      }

    }
  }

  &--handleActive{
    padding-left: calc(#{simple()} + .5rem);
    cursor: pointer;
  }
}
