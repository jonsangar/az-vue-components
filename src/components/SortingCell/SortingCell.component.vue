<template lang="pug">
.az-sorting-cell(
  :class="getClass"
  @click="onClick")

  az-icon.handleIcon(
    v-if="activeHandle"
    name="order" scale="1.5")
  //- .handle(
  //-   v-if="!settingsActive")

  //- .handle(
  //-   v-if="!isHeader && activeHandle")
  //-   az-icon(name="order" scale="1.5")

  //- .handle--header(
  //-   v-if="isHeader && activeHandle")
  //-   az-icon(name="order" scale="1.5")

  .az-sorting-cell__content(
    v-if="!$slots.custom")
    .az-sorting-cell__content__title(
      v-if="!settingsActive || !editMode"
      @click="activeEditMode")
      | {{ localName }}

    .az-sorting-cell__content__title(
      v-show="settingsActive && editMode")
      form(@submit.prevent="onSubmit")
        input.az-sorting-cell__content__title--input(
          :id="`input-item-${itemID}`"
          :disabled="!editMode"
          ref="name"
          :maxlength="maxlength"
          :minlength="minlength"
          @keyup.enter="onEnter"
          @blur="onBlur"
          v-model="localName")

    .az-sorting-cell__content__subtitle(
      v-if="validation")
      | {{validation}}
  .az-sorting-cell__content(
    v-else)
    slot(name="custom")

    .az-sorting-cell__content__subtitle(
      v-if="validation")
      | {{validation}}




  //- .editButton(
  //-   v-if="settingsActive && !editMode"
  //-   @click="activeEditMode")
  //-   az-icon.az-icon--linght(name="edit2" scale="1")

  //- .editButton(
  //-   v-if="settingsActive && editMode"
  //-   @click="saveChanges")
  //-   az-icon.az-icon--success(name="save" scale="1")

</template>

<script>
export default {

  props: {
    name: {
      type: String,
      required: false
    },

    validation: {
      type: String,
      default: ''
    },

    itemID: {
      type: [String, Number],
      required: false
    },

    settingsActive: {
      type: Boolean,
      default: false
    },

    activeHandle: {
      type: Boolean,
      default: true
    },

    isHeader: {
      type: Boolean,
      default: false
    },

    maxlength: {
      type: Number,
      required: false,
    },

    minlength: {
      type: Number,
      required: false,
    },

    keyboardAdjust: {
      type: Boolean,
      default: false
    }

  },

  data() {
    return {
      localName: this.name,
      editMode: false,
      debounceActive: false,
      isAndroid: this.$constants.IS_MOBILE === 'android'
    }
  },

  watch: {
    name(newValue){
      this.localName = newValue;
    },
  },

  computed:{
    getClass() {
      return [{ 'handle': this.activeHandle === true || this.$slots.custom,
                'az-sorting-cell--settingsActive': this.settingsActive === true,
                'az-sorting-cell--header': this.isHeader === true,
                'az-sorting-cell--handleActive': this.activeHandle === true}]
    }
  },

  methods: {
    initialState() {

      if(this.isAndroid && this.keyboardAdjust && this.itemID) {
        const cell = document.getElementById(this.itemID);
        const input = document.getElementById(`input-item-${this.itemID}`);
        // El contenedor sobre el que hacer scroll, si no es main (caso genérico), será
        //  main-override (en vista InstallationSort) o "sortingZones", en
        // la vista de ordenar zonas.
        const container = document.getElementsByClassName("main")[0] ?
          document.getElementsByClassName("main")[0] :
          (document.getElementsByClassName("sortingZones")[0] ?
            document.getElementsByClassName("sortingZones")[0] :
            document.getElementsByClassName("main-override")[0]);


        input.addEventListener('focus', () => {
          document.body.classList.add('keyboard-up');
          // Si la celda está por debajo de la mitad de la pantalla, entonces
          // hacemos el scroll del teclado
          if(cell.getBoundingClientRect().top >= container.clientHeight / 2) {
            // setTimeout(() => {
              if(container !== null){
                container.scrollTo(0, container.scrollHeight);
              }
            // }, 100)
          }
        });

        input.addEventListener('blur', () => {
          // setTimeout(() => {
            document.body.classList.remove('keyboard-up')
          // }, 500);
        });



      }
    },

    activeEditMode() {
      if(!this.settingsActive) return;

      this.editMode = true;
      // Next Tick nos asegura que se ha habilitado el campos antes de tomar el foco
      this.$nextTick(() => this.$refs.name.focus());

      this.$emit('focus');
    },

    onSubmit() {
      this.saveChanges();
    },
    // Manejamos blur del formulario de nombre para guardar datos
    onBlur(){
      this.onSubmit();
    },
    // Manejamos tecla enter del formulario de nombre para guardar datos
    onEnter(){
      this.$refs.name.blur();
    },

    onClick(e) {
      this.$emit('click', e);
    },

    saveChanges() {
      // Aplicamos un pequeño retardo para evitar rebotes.
      // Al guardar esperamos 200ms antes de permitir activar "activeEditMode"
      // setTimeout(() => {
      //   this.debounceActive = false;
      // }, 200)
      // this.editMode = false;
      if(this.name === this.localName){
        this.editMode = false;
        this.$emit('save', null);
        return;
      }
      // console.log("Guardando", this.localName);

      // Incluímos un pequeño retardo para que de tiempo a validar
      // desde el componente padre antes de desactivar la edición
      this.$emit('save', {id: this.itemID, name: this.localName});

      setTimeout(() => {
        if(!this.validation){
          this.editMode = false;
        }
      }, 500)
    }
  },

  mounted() {
    this.$nextTick(() => this.initialState());
  }
}
</script>

<style lang="scss" src="./SortingCell.component.scss"></style>
