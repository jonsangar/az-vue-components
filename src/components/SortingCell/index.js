import SortingCell from './SortingCell.component.vue';

export default SortingCell;
