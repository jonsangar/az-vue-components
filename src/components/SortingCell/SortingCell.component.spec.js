import { shallowMount, createLocalVue } from '@vue/test-utils';
import { fireEvent } from '@testing-library/vue';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import SortingCell from './SortingCell.component.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

const elementMock = { addEventListener: jest.fn() };
jest.spyOn(document, 'getElementById').mockImplementation(() => elementMock);

describe('SortingCell.vue', () => {

  const build = () => {
    const wrapper = shallowMount(SortingCell, {
      localVue,
      propsData: {
        name: 'MyTitle',
        itemID: 'MyItem',
        settingsActive: true,
        activeHandle: false,
        isHeader: true,
        maxlength: 10,
        minlength: 2,
      },
      mocks: {
        $constants: {
          IS_MOBILE: 'android'
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente SortingCell con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Comprueba que se actualizan las props cuando se asignan las props', async () => {
    const { wrapper } = build()

    expect(wrapper.vm.localName).toBe('MyTitle')
    expect(wrapper.vm.itemID).toBe('MyItem')
    expect(wrapper.vm.settingsActive).toBe(true)
    expect(wrapper.vm.activeHandle).toBe(false)
    expect(wrapper.vm.isHeader).toBe(true)
    expect(wrapper.vm.maxlength).toBe(10)
    expect(wrapper.vm.minlength).toBe(2)
  });

  it('Comprueba que se renderiza las props cuando se asignan las props', async () => {
    const { wrapper } = build()

    expect(wrapper.find('.az-sorting-cell__content__title').text()).toBe('MyTitle')
    expect(wrapper.find('#input-item-MyItem').exists()).toBe(true)
    expect(wrapper.find('#input-item-MyItem').attributes().disabled).toBe('disabled')
  });
})

describe('SortingCell.vue', () => {

  const build = () => {
    const wrapper = shallowMount(SortingCell, {
      localVue,
      propsData: {
        name: 'MyTitle',
        itemID: 'MyItem',
        settingsActive: true,
      },
      mocks: {
        $constants: {
          IS_MOBILE: 'android'
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que emite el evento focus cuando se activa el modo edición', async () => {
    const { wrapper } = build()

    await wrapper.vm.activeEditMode()
    expect(wrapper.emitted().focus).toBeTruthy()
    expect(wrapper.emitted().focus.length).toBe(1)
    expect(wrapper.emitted().focus[0]).toEqual([])
  });

  it('Comprueba que emite el evento save cuando se ejecuta onSubmit', async () => {
    const { wrapper } = build()

    await wrapper.vm.onSubmit()
    expect(wrapper.emitted().save).toBeTruthy()
    expect(wrapper.emitted().save.length).toBe(1)
    expect(wrapper.emitted().save[0]).toEqual([null])
  });

  it('Comprueba que emite el evento save cuando se hace blur en el input', async () => {
    const { wrapper } = build()

    // Tomamos el foco del input
    const input = wrapper.find('input');
    // Simulamos tomar el foco sobre el input
    await fireEvent.blur(input.element);
    expect(wrapper.emitted().save).toBeTruthy()
    expect(wrapper.emitted().save.length).toBe(1)
    expect(wrapper.emitted().save[0]).toEqual([null])
  });

  it('Comprueba que emite el evento save cuando se pulsa enter en el input', async () => {
    const { wrapper } = build()

    await wrapper.vm.activeEditMode()
    // Tomamos el foco del input
    // const input = wrapper.find('input');
    // Simulamos tomar el foco sobre el input
    // await fireEvent.keyUp(input.element);
    await wrapper.vm.onEnter()
    // console.log(wrapper.vm.$refs.name)
    await wrapper.vm.$refs.name.blur()
    expect(wrapper.vm.$refs.name.value).toBe('MyTitle')
    await wrapper.vm.onBlur()
    expect(wrapper.emitted().save).toBeTruthy()
    expect(wrapper.emitted().save.length).toBe(1)
    expect(wrapper.emitted().save[0]).toEqual([null])
  });
})

describe('SortingCell.vue', () => {

  const build = () => {
    const wrapper = shallowMount(SortingCell, {
      localVue,
      propsData: {
         name: 'MyTitle',
        itemID: 'MyItem',
        settingsActive: true,
        keyboardAdjust: true,
        isAndroid: true,
      },
      mocks: {
        $constants: {
          IS_MOBILE: 'android'
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que emite el evento save cuando se hace blur en el input y el valor de localName es distinto del valor de la prop name', async () => {
    // fake timers
    jest.useFakeTimers()

    const { wrapper } = build()

    // Asignamos un valor distinto a localName para que guarde itemID y localName
    wrapper.vm.localName = 'MyCustomName'
    // Tomamos el foco del input
    const input = wrapper.find('input');
    // Simulamos tomar el foco sobre el input
    await fireEvent.blur(input.element);

    // Wait to setTimout 500
    jest.runTimersToTime(510)

    expect(wrapper.vm.editMode).toBe(false)
    expect(wrapper.emitted().save).toBeTruthy()
    expect(wrapper.emitted().save.length).toBe(1)
    expect(wrapper.emitted().save[0]).toEqual([{id: wrapper.vm.itemID, name: wrapper.vm.localName}])
  });
});
