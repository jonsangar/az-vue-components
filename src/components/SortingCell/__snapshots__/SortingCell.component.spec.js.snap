// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortingCell.vue Comprueba renderizado del componente SortingCell con snapshot 1`] = `
<div class="az-sorting-cell az-sorting-cell--settingsActive az-sorting-cell--header">
  <!---->
  <div class="az-sorting-cell__content">
    <div class="az-sorting-cell__content__title">MyTitle</div>
    <div class="az-sorting-cell__content__title" style="display: none;">
      <form><input id="input-item-MyItem" disabled="disabled" maxlength="10" minlength="2" class="az-sorting-cell__content__title--input"></form>
    </div>
    <!---->
  </div>
</div>
`;
